packages/ContentAutotagging/Engine/generic/AutotagBaseEngine.ts

```diff
@@ -0,0 +1,745 @@
+        await contentItem.Delete()
+    }
+
+    public chunkExtractedText(text: string, modelAPIName: string): string[]{
+        try {
+            const tokenLimit = ModelTokenLimits[modelAPIName]
+            const textLimit = Math.ceil(tokenLimit / 1.5 ) // bit of a conservatice estimate to ensure there is room for the additional prompts
+            
+            if (text.length <= textLimit) {
+                // No need to chunk the text
+                return [text]
+            }
+
+            const numChunks = Math.ceil(text.length / textLimit)
+            const chunkSize = Math.ceil(text.length / numChunks)
+            const chunks = []
+            for (let i = 0; i < numChunks; i++) {
+                const start = i * chunkSize
+                const end = (i + 1) * chunkSize
+                chunks.push(text.slice(start, end))
+            }
+            return chunks
+        } catch (e) {
+            console.log('Could not chunk the text')
+            return [text]
+        }
+    }
+
+    /**
+     * Given the processing results from the LLM and the Content Element Item that was saved to the database, this function saves the tags as Content Element Tags in the database.
+     * @param md: The metadata object to save the tags
+     * @param contentElementItem: The content element item that was saved to the database
+     * @param results: The results of the processing from the LLM
+     * @param contextUser: The user context to save the tags
+     * @returns
+     */
+    public async saveContentItemTags(contentItemID: string, LLMResults: JsonObject, contextUser: UserInfo) {
+        const md = new Metadata()
+        for (const keyword of LLMResults.keywords) {
+
+            const contentItemTags = <ContentItemTagEntity> await md.GetEntityObject('Content Item Tags', contextUser)
+            
+            contentItemTags.NewRecord()
+            contentItemTags.Set('ItemID', contentItemID)
+            contentItemTags.Set('Tag', keyword)
+            contentItemTags.Set('CreatedAt', LLMResults.processEndTime)
+            contentItemTags.Set('UpdatedAt', LLMResults.processEndTime)
+
+            await contentItemTags.Save()
+        }
+    }
+
+    public async saveResultsToContentItemAttribute(LLMResults: JsonObject, contextUser: UserInfo) {
+        const md = new Metadata()
+        for (const key in LLMResults) {
+            if (key !== 'keywords' && key !== 'processStartTime' && key !== 'processEndTime' && key !== 'contentItemID' && key !== 'isValidContent') {
+                const contentItemAttributes = <ContentItemAttributeEntity> await md.GetEntityObject('Content Item Attributes', contextUser)
+                
+                //Value should be a string, if its a null or undefined value, set it to an empty string
+                const value = LLMResults[key] || ''
+
+                contentItemAttributes.NewRecord()
+                contentItemAttributes.Set('ContentItemID', LLMResults.contentItemID)
```

Comment:
Why weakly typed use of `BaseEntity` instead of using generated strongly-typed sub-classes @nico-oz

Reply:
Was it a circular dep issue?